docs/common-problems.asciidoc

```diff
@@ -240,3 +241,29 @@ For example:
 APM Agent --> Load Balancer  --> APM Server 
    10s            15s               30s
 ----------------------------------------------------------------------
+
+[[server-es-down]]
+[float]
+=== What happens to data when APM Server or Elasticsearch is down?
+
+==== Elasticsearch down
+
+If Elasticsearch goes down, the APM Server will keep data in memory until Elasticsearch is back up,
+or until it runs out of space in its internal in-memory queue. 
+You can <<adjust-queue-size,adjust the internal queue size>> if necessary.
+When the queue becomes full, APM Server will respond with an <<queue-full`HTTP 503: Queue is full`>>,
+and data will be lost.
+
+==== APM Server down
+
+Some agents have internal queues or buffers to temporarily store data in case the APM Server goes down.
+As a general rule of thumb, queues fill up quickly. Assume data will be lost if APM Server goes down.
+Adjusting these queues/buffers can increase the overhead of the agent, so use caution when updating default values.
+
+* Go Agent - `ELASTIC_APM_BUFFER_SIZE`/`ELASTIC_APM_REQUEST_SIZE`?
```

Comment:
@axw - can you help me understand these settings in this context?

Reply:
It's a bit complicated :)

https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-api-buffer-size
https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-api-request-size

There's actually four levels of queuing/buffering:
 - a queue (channel in Go parlance) between the application code and the background "tracer" goroutine. The size of this queue is not configurable, but data will never stay on there for long anyway
 - a circular buffer whose size is controlled by `ELASTIC_APM_BUFFER_SIZE`. When the tracer receives an event from the channel, it will add it to the buffer. If the buffer is full, then old events will be dropped until the new event can be added
 - a zlib compression buffer. The size is not configurable, data just goes into the buffer until there's enough to compress, or is flushed to the request buffer
 - a request buffer, which holds the ND-JSON encoded and zlib compressed payload. The size is configured with `ELASTIC_APM_REQUEST_SIZE`

Increasing `ELASTIC_APM_BUFFER_SIZE` means that more events can be kept in memory, e.g. because the server is unreachable, or because the application is sending events at a faster rate than the agent can send them to the server. 

Increasing `ELASTIC_APM_REQUEST_SIZE` means that more events may be sent to the APM Server in a single request. I think it would be rare to need to change this.

Reply:
oooo that is complicated. Thanks for the thorough explanation!

If I'm understanding correctly, then it makes sense to me to mention the circular buffer and its config variable: `ELASTIC_APM_BUFFER_SIZE` here then.